python/ray/autoscaler/autoscaler.py

```diff
@@ -0,0 +1,329 @@
+# is a safety feature to prevent e.g. runaway node launches.
+MAX_NUM_FAILURES = 5
+
+# Max number of nodes to launch at a time.
+MAX_CONCURRENT_LAUNCHES = 10
+
+
+class StandardAutoscaler(object):
+    """The autoscaling control loop for a Ray cluster.
+
+    There are two ways to start an autoscaling cluster: manually by running
+    `ray start --head --autoscaling-config=/path/to/config.json` on a
+    instance that has permission to launch other instances, or you can also use
+    `ray bootstrap /path/to/config.json` from your laptop, which will configure
+    the right AWS/Cloud roles automatically.
+
+    StandardAutoscaler's `update` method is periodically called by `monitor.py`
+    to add and remove nodes as necessary. Currently, load-based autoscaling is
+    not implemented, so all this class does is try to maintain a constant
+    cluster size.
+
+    StandardAutoscaler is also used to bootstrap clusters (by adding workers
+    until the target cluster size is met).
+    """
+
+    def __init__(
+            self, config_path,
+            max_concurrent_launches=MAX_CONCURRENT_LAUNCHES,
+            max_failures=MAX_NUM_FAILURES, process_runner=subprocess,
+            verbose_updates=False, node_updater_cls=NodeUpdaterProcess):
+        self.config_path = config_path
+        self.reload_config(errors_fatal=True)
+        self.provider = get_node_provider(
+            self.config["provider"], self.config["cluster_name"])
+
+        self.max_failures = max_failures
+        self.max_concurrent_launches = max_concurrent_launches
+        self.verbose_updates = verbose_updates
+        self.process_runner = process_runner
+        self.node_updater_cls = node_updater_cls
+
+        # Map from node_id to NodeUpdater processes
+        self.updaters = {}
+        self.num_failed_updates = defaultdict(int)
+        self.num_failures = 0
+
+        for local_path in self.config["file_mounts"].values():
+            assert os.path.exists(local_path)
+
+        print("StandardAutoscaler: {}".format(self.config))
+
+    def update(self):
+        try:
+            self.reload_config(errors_fatal=False)
+            self._update()
+        except Exception as e:
+            print(
+                "StandardAutoscaler: Error during autoscaling: {}",
+                traceback.format_exc())
+            self.num_failures += 1
+            if self.num_failures > self.max_failures:
+                print("*** StandardAutoscaler: Too many errors, abort. ***")
+                raise e
+
+    def _update(self):
+        nodes = self.workers()
+        target_num_workers = self.config["max_workers"]
+
+        # Terminate nodes while there are too many
+        while len(nodes) > target_num_workers:
+            print(
+                "StandardAutoscaler: Terminating unneeded node: "
+                "{}".format(nodes[-1]))
+            self.provider.terminate_node(nodes[-1])
```

Comment:
naive question - Is this ok? will this have weird effects on fault tolerance/do you need to drain?

Reply:
I'm guessing the fault tolerance will not be very happy with this since we don't have task checkpoints. That's fine for now, GCS integration is not implemented.

Reply:
@richardliaw what do you mean by "drain"? Do you mean transfer objects away from that node to other nodes?

Reply:
Oh I just meant letting tasks finish while not scheduling any more tasks on that node